packages/shopping-cart/src/shopping-cart-manager.ts

```diff
@@ -0,0 +1,204 @@
+	createActionPromisesManager,
+	noopManager,
+} from './managers';
+import { createActionCreators } from './shopping-cart-actions';
+import { getInitialShoppingCartState, shoppingCartReducer } from './shopping-cart-reducer';
+import { createTakeActionsBasedOnState } from './state-based-actions';
+import { createCartSyncMiddleware, createCartInitMiddleware } from './sync';
+import type {
+	GetCart,
+	SetCart,
+	ShoppingCartManagerClient,
+	ShoppingCartManagerState,
+	ShoppingCartManager,
+	RequestCart,
+	ShoppingCartAction,
+	ResponseCart,
+	ShoppingCartReducerDispatch,
+	DispatchAndWaitForValid,
+	AddActionPromise,
+	ShoppingCartState,
+	UseShoppingCart,
+} from './types';
+
+const debug = debugFactory( 'shopping-cart:shopping-cart-manager' );
+const emptyCart = getEmptyResponseCart();
+const getEmptyCart = () => Promise.resolve( emptyCart );
+
+function createDispatchAndWaitForValid(
+	dispatch: ShoppingCartReducerDispatch,
+	addActionPromise: AddActionPromise
+): DispatchAndWaitForValid {
+	const dispatchAndWaitForValid: DispatchAndWaitForValid = ( action ) => {
+		return new Promise< ResponseCart >( ( resolve ) => {
+			dispatch( action );
+			addActionPromise( resolve );
+		} );
+	};
+
+	return dispatchAndWaitForValid;
+}
+
+function createShoppingCartManager(
+	cartKey: string,
+	getCart: GetCart,
+	setCart: SetCart
+): ShoppingCartManager {
+	let state = getInitialShoppingCartState();
+
+	const shouldNotFetchRealCart = cartKeysThatDoNotAllowFetch.includes( cartKey );
+
+	const setServerCart = ( cartParam: RequestCart ) => setCart( cartKey, cartParam );
+	const getServerCart = () => {
+		if ( shouldNotFetchRealCart ) {
+			return getEmptyCart();
+		}
+		return getCart( cartKey );
+	};
+
+	const syncCartToServer = createCartSyncMiddleware( setServerCart );
+	const initializeCartFromServer = createCartInitMiddleware( getServerCart );
+	const middleware = [ initializeCartFromServer, syncCartToServer ];
+
+	const { subscribe, notifySubscribers } = createSubscriptionManager( cartKey );
+
+	// When an action is dispatched that modifies the cart (eg:
+	// addProductsToCart), the reducer modifies the `responseCart` stored in
+	// state. That (incomplete) data is then sent off to the server to be
+	// validated and filled-in with additional properties before being returned
+	// to the ShoppingCartManager. Because of this, optimistic updating of the
+	// cart is not possible (it may change significantly in that round trip) so
+	// we don't want to return the raw `responseCart` to the consumer. Instead,
+	// we keep a copy of the `responseCart` the last time the state had a `valid`
+	// CacheStatus and pass that to our consumers. The consumers can use
+	// `isPendingUpdate` to know when the cart data is updating.
+	const {
+		getLastValidResponseCart,
+		updateLastValidResponseCart,
+	} = createLastValidResponseCartManager( state );
+
+	const { resolveActionPromisesIfValid, addActionPromise } = createActionPromisesManager();
```

Comment:
The `LastValidResponseCartManager`, `ActionPromisesManager` and `SubscriptionManager` are typical objects. They have some internal state and methods to manipulate it. Therefore, I think it would be more readable if we used the OOP syntax for them, like:
```js
const lastValidCart = createLastValidCart( initialState );
lastValidCart.get();
lastValidCart.update( cart );
```
The method names can be now much shorter, just a verb, because the subject is in the object name.

And if something depends on the object, we pass the entire object instead of just a method:
```js
createTakeActionBasedOnState( lastValidCart, actionPromises );
```